src/librustc/ty/mod.rs

```diff
@@ -2891,7 +2901,26 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn impls_are_allowed_to_overlap(self, def_id1: DefId, def_id2: DefId)
                                         -> Option<ImplOverlapKind>
     {
-        let is_legit = if self.features().overlapping_marker_traits {
+        match (self.impl_polarity(def_id1), self.impl_polarity(def_id2)) {
+            (ImplPolarity::Reservation, _) |
+            (_, ImplPolarity::Reservation) => {
+                // `#[rustc_reservation_impl]` impls don't overlap with anything
+                debug!("impls_are_allowed_to_overlap({:?}, {:?}) = Some(Permitted) (reservations)",
+                       def_id1, def_id2);
+                return Some(ImplOverlapKind::Permitted);
+            }
+            (ImplPolarity::Positive, ImplPolarity::Negative) |
+            (ImplPolarity::Negative, ImplPolarity::Positive) => {
+                // FIXME: when can this happen?
```

Comment:
Isn't that this case? ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5a8191e63a5348bd32d52f675f1be8fc))

```rust
#![feature(optin_builtin_traits)]

auto trait Foo { }

impl Foo for u32 { }
impl !Foo for u32 { }

fn main() { }
```